File: csharp/HearthstoneReplays/ReplayData/Entities/BaseEntity.cs
#region

using System.Collections.Generic;
using System.Linq;
using System.Xml.Serialization;
using HearthstoneReplays.ReplayData.GameActions;

#endregion

namespace HearthstoneReplays.ReplayData.Entities
{
	[XmlInclude(typeof(GameEntity))]
	[XmlInclude(typeof(PlayerEntity))]
	[XmlInclude(typeof(FullEntity))]
	public abstract class BaseEntity : GameData
	{
		[XmlAttribute("id")]
		public int Id { get; set; }

		[XmlElement("Tag", typeof(Tag))]
		public List<Tag> Tags { get; set; }

		public override bool Equals(object obj)
		{
			var other = obj as BaseEntity;
			if(other == null)
				return false;
			return Id == other.Id && Tags.All(tag => other.Tags.Any(t1 => t1.Name == tag.Name && t1.Value == tag.Value));
		}
	}
}

File: java/src/main/java/com/hearthsim/hsreplay/parser/replaydata/entities/BaseEntity.java
package com.hearthsim.hsreplay.parser.replaydata.entities;

import java.util.ArrayList;
import java.util.List;

import javax.xml.bind.annotation.XmlAccessOrder;
import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorOrder;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;

import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

import com.hearthsim.hsreplay.parser.replaydata.GameData;
import com.hearthsim.hsreplay.parser.replaydata.gameactions.Tag;

@Getter
@Setter
@EqualsAndHashCode(of = { "id", "tags" }, callSuper = false)
@AllArgsConstructor
@NoArgsConstructor
@XmlAccessorType(XmlAccessType.FIELD)
@ToString
@XmlAccessorOrder(XmlAccessOrder.ALPHABETICAL)
public abstract class BaseEntity extends GameData {

	@XmlAttribute(name = "id")
	protected int id;

	@XmlElement(name = "Tag")
	protected List<Tag> tags = new ArrayList<>();
}